### DISCUSSION TITLE
Feature Request: expose client property on withTRPC HOC to assign instance of createTRPCClient()

### DISCUSSION THREAD

**Discussion by uncvrd** [trpc/trpc#1854] [Ideas]
Hi! I had a brief discussion on the trpc Discord and I'd like to suggest the following modification to the `withTRPC()` HOC.

There are scenarios where a user may need to access the `trpc` client outside of the react context. For example, when using `xstate` for state management, you can provide methods to run "side effects" inside your definitions. However, `xstate` defines its processes in JS outside of the react context (as a result I cannot use the `trpc` hooks for querying/mutating). Therefore, if I wish to use `trpc` with NextJS and XState, I would have to create a client twice:

1. Once inside the `withTRPC` where the return statement defines the `createTRPCClient` config properties.
2. Second, I would have to define a `@trpc/client` raw JS implementation of `createTRPCClient()` to use in my XState files.

Therefore, I would like to propose the following change to `withTRPC` which would allow assignment of a single client app-wide:

```
// lib/trpc.ts
// define a single client
// now I can pass this instance around anywhere inside or outside of React!
export const trpcClient = createTRPCClient<AppRouter>({
    links: [
        httpBatchLink({
            url: `${getBaseUrl()}/api/trpc`,
        }),
    ],
})
```

```
//_app.ts
import { trpcClient } from "~/lib/trpc"

// pass trpcClient in to the configuration...maybe even return trpcClient directly instead of inside the return object?
export default withTRPC<AppRouter>({
  config(info) {
    return {
      client: trpcClient
    };
  },
})(MyApp);
```

Thoughts?

This was also discussed briefly here: https://github.com/trpc/trpc/discussions/1351#discussion-3752219

**Comment by KATT** [comment]
This was actually how it first worked, but then when SSRing the instance will be shared across *all* users which is obviously.... bad.

For now, did it work to write it to a local variable on mount?

**Reply by uncvrd** [reply]
Oh yikes! That's a fair point and not good.

The local variable worked - but it definitely felt like a hack and I wasn't too comfortable with the implementation.

I got the idea for defining the instance manually from this comment at the `xstate` repository referring to `react-query` sharing an instance but I didn't take in to account SSR before posting. https://github.com/statelyai/xstate/discussions/1813#discussioncomment-262572

I think it would be nice to be able to do "either or" but that sounds like a nightmare to handle for conditionals in the config and making sure developers understand the implications of each implementation.

**Reply by KATT** [reply]
I rather have this setup where its less likely to do fuckups. The hack is alright imo. Don't xstate have a way of using it in a hook?

**Reply by KATT** [reply]
Surely the same challenge should occur in graphql/apllo/urql land

**Reply by uncvrd** [reply]
Agreed - and yea, xstate does allow you to define queries within context using a `useMachine()` hook, but it gets a bit complicated doing so with dynamic nested machines spawning actors and those actors are the ones that I need to write `trpc` requests for (I can expand further but not relevant, it's just something I haven't found a way to do haha). I was hoping to be able to define my query implementations directly within my statemachine outside react without having to take the hooks route 😄 

No worries though - I appreciate the insight and will stick with the workaround for now
